src/cds/src/main/kotlin/org/icpclive/cds/cats/CATSDataSource.kt

```diff
@@ -0,0 +1,195 @@
+package org.icpclive.cds.cats
+
+import kotlinx.coroutines.*
+import kotlinx.datetime.Clock
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import org.icpclive.api.*
+import org.icpclive.cds.ContestParseResult
+import org.icpclive.cds.FullReloadContestDataSource
+import org.icpclive.cds.common.jsonLoaderService
+import org.icpclive.util.getLogger
+import java.awt.Color
+import java.util.*
+import kotlin.time.Duration.Companion.hours
+import kotlin.time.Duration.Companion.seconds
+
+
+class CATSDataSource(val properties: Properties, creds: Map<String, String>) : FullReloadContestDataSource(5.seconds) {
+    private val login = creds["cats.login"]
+    private val password = creds["cats.password"]
+    private val url = properties.getProperty("url")
+
+    private val logger = getLogger(CATSDataSource::class)
+
+    // :TODO I may send less count of quires to the server but I need to know when my sid is expired
+    private fun sid(): String = runBlocking {
+        async { authLoader.loadOnce().sid }.await()
+    }
+
+    @Serializable
+    data class Auth(val status: String, val sid: String, val cid: Long)
+
+    @Serializable
+    data class Problem(val id: Int, val name: String, val code: String)
+
+    @Serializable
+    data class Problems(val problems: List<Problem>)
+
+    @Serializable
+    data class Team(val id: Int, val account_id: Int, val login: String, val name: String, val role: String)
+
+    @Serializable
+    data class Users(val users: List<Team>)
+
+    @Serializable
+    data class Contest(
+        val title: String,
+        val start_date: String,
+        val freeze_date: String,
+        val finish_date: String,
+        val rules: String
+    )
+
+    @Serializable
+    sealed class Run
+    
+    @Serializable
+    @SerialName("submit")
+    data class Submit(
+        val id: Int,
+        val state_text: String,
+        val problem_id: Int,
+        val team_id: Int,
+        val time: String
+        // ::TODO add failed tests
+    ) : Run()
+
```

Comment:
So cool, it's working!
But I don't understand why. So when we write SerialName("text") it tries to find "text" in values of all json objects with key "type"?
And second question, I fix it but I worry about _strange quiries_ for example you may see that I add a lot of stuff classes but if author of api cats decided to add new type then my program throw exception:(

Reply:
Basically it sais, that Run can have several different types. As Run is sealed class, all its subclasses must be here, so you can switch all of them. So when you deserialize Run object it uses some strategy to choose which of its (known in advance list) subtypes to deserialize.  

By default this strategy is check field "type" and choose by name of it's class. There are some other [options](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/polymorphism.md#closed-polymorphism)

Reply:
It will fail, if there is a new type of run. I'm not sure if it's bad. Ignoring is quite easy anyway. The same will happen, if there is something changed in normal run object, and it will fail to parse. Maybe it's something significant, who knows?

But to be honest, sending clarifications in run API is really a strange idea.